File: frontend/types/chart.ts
export type ChartType =
  | "line"
  | "bar"
  | "scatter"
  | "area"
  | "table";


export type ChartValue =
  | string
  | number
  | boolean
  | null;


export type ChartDataRow =
  Record<string, ChartValue>;


export interface ChartRequest {
  dataset_id: string;

  chart_type: ChartType;

  x?: string | null;

  series?: string[];

  title?: string | null;

  max_rows?: number;
}


export interface ChartSpec {
  ok: boolean;

  dataset_id: string;
  file_name: string;

  chart_type: ChartType;

  title: string;

  x?: string | null;

  series: string[];

  data: ChartDataRow[];

  source_row_count: number;
  chart_row_count: number;

  sampled: boolean;

  explanation: string;

  source_coverage: number;

  warnings: string[];
}


export interface ChartApiResponse
  extends ChartSpec {
  requestId?: string;

  error?: string;
  detail?: string;
}

File: agent-service/app/schemas/chart.py
from typing import Any, Literal

from pydantic import BaseModel, Field


ChartType = Literal[
    "line",
    "bar",
    "scatter",
    "area",
    "table",
]


class ChartRequest(BaseModel):
    dataset_id: str = Field(
        min_length=1,
    )

    chart_type: ChartType

    x: str | None = None

    series: list[str] = Field(
        default_factory=list,
    )

    title: str | None = None

    max_rows: int = Field(
        default=500,
        ge=10,
        le=2000,
    )


class ChartSpecResponse(BaseModel):
    ok: bool = True

    dataset_id: str
    file_name: str

    chart_type: ChartType

    title: str

    x: str | None = None

    series: list[str] = Field(
        default_factory=list,
    )

    data: list[
        dict[str, Any]
    ] = Field(
        default_factory=list,
    )

    source_row_count: int
    chart_row_count: int

    sampled: bool = False

    explanation: str

    source_coverage: float = Field(
        default=1.0,
        ge=0.0,
        le=1.0,
    )

    warnings: list[str] = Field(
        default_factory=list,
    )